src/shell/output.c

```diff
@@ -345,6 +345,47 @@ static int shell_output_write_fd (int fd, const void *buf, size_t len)
     return n;
 }
 
+static int shell_output_label (struct shell_output_type_file *ofp,
+                               const char *rank)
+{
+    if (shell_output_write_fd (ofp->fdp->fd,  rank, strlen (rank)) < 0
+        || shell_output_write_fd (ofp->fdp->fd, ": ", 2) < 0)
+        return -1;
+    return 0;
+}
+
+static int shell_output_data (struct shell_output *out, json_t *context)
+{
+    struct shell_output_type_file *ofp;
+    int output_type;
+    const char *stream = NULL;
+    const char *rank = NULL;
+    char *data = NULL;
+    int len = 0;
+
+    if (iodecode (context, &stream, &rank, &data, &len, NULL) < 0) {
+        shell_log_errno ("iodecode");
+        return -1;
+    }
+    if (!strcmp (stream, "stdout")) {
+        output_type = out->stdout_type;
+        ofp = &out->stdout_file;
+    }
+    else {
+        output_type = out->stderr_type;
+        ofp = &out->stderr_file;
+    }
+    if ((output_type == FLUX_OUTPUT_TYPE_FILE) && len > 0) {
+        if (ofp->label
+            && shell_output_label (ofp, rank) < 0)
+            return -1;
+        if (shell_output_write_fd (ofp->fdp->fd, data, len) < 0)
+            return -1;
+    }
```

Comment:
Is `data` leaked on error?

Reply:
yes, good catch!

Reply:
I was going to add a separate commit for this fix since it was present in existing code, however I just squashed it into the refactor commit. Hope that was ok!